source/extensions/filters/listener/http_inspector/http_inspector.cc

```diff
@@ -0,0 +1,174 @@
+#include "extensions/filters/listener/http_inspector/http_inspector.h"
+
+#include "envoy/event/dispatcher.h"
+#include "envoy/network/listen_socket.h"
+#include "envoy/stats/scope.h"
+
+#include "common/api/os_sys_calls_impl.h"
+#include "common/common/assert.h"
+#include "common/common/macros.h"
+#include "common/http/headers.h"
+
+#include "extensions/transport_sockets/well_known_names.h"
+
+#include "absl/strings/match.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace ListenerFilters {
+namespace HttpInspector {
+
+Config::Config(Stats::Scope& scope)
+    : stats_{ALL_HTTP_INSPECTOR_STATS(POOL_COUNTER_PREFIX(scope, "http_inspector."))} {}
+
+const absl::string_view Filter::HTTP2_CONNECTION_PREFACE = "PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n";
+thread_local uint8_t Filter::buf_[Config::MAX_INSPECT_SIZE];
+
+Filter::Filter(const ConfigSharedPtr config) : config_(config) {}
+
+Network::FilterStatus Filter::onAccept(Network::ListenerFilterCallbacks& cb) {
+  ENVOY_LOG(debug, "http inspector: new connection accepted");
+
+  Network::ConnectionSocket& socket = cb.socket();
+
+  absl::string_view transport_protocol = socket.detectedTransportProtocol();
+  if (!transport_protocol.empty() &&
+      transport_protocol != TransportSockets::TransportSocketNames::get().RawBuffer) {
+    ENVOY_LOG(trace, "http inspector: cannot inspect http protocol with transport socket {}",
+              transport_protocol);
+    return Network::FilterStatus::Continue;
+  }
+
+  ASSERT(file_event_ == nullptr);
+
+  file_event_ = cb.dispatcher().createFileEvent(
+      socket.ioHandle().fd(),
+      [this](uint32_t events) {
+        ASSERT(events == Event::FileReadyType::Read);
+        onRead();
+      },
+      Event::FileTriggerType::Edge, Event::FileReadyType::Read);
+
+  cb_ = &cb;
+  return Network::FilterStatus::StopIteration;
+}
+
+void Filter::onRead() {
+  auto& os_syscalls = Api::OsSysCallsSingleton::get();
+  const Api::SysCallSizeResult result =
+      os_syscalls.recv(cb_->socket().ioHandle().fd(), buf_, Config::MAX_INSPECT_SIZE, MSG_PEEK);
+  ENVOY_LOG(trace, "http inspector: recv: {}", result.rc_);
+
+  if (result.rc_ == -1 && result.errno_ == EAGAIN) {
+    return;
+  } else if (result.rc_ < 0) {
+    config_->stats().read_error_.inc();
+    done(false);
+    return;
+  }
+
+  if (static_cast<uint64_t>(result.rc_) > read_) {
+    const uint8_t* data = buf_ + read_;
+    const size_t len = result.rc_ - read_;
+    read_ = result.rc_;
+    parseHttpHeader(absl::string_view(reinterpret_cast<const char*>(data), len));
+  } else {
+    done(false);
+  }
+}
+
+void Filter::parseHttpHeader(absl::string_view data) {
+  if (absl::StartsWith(data, Filter::HTTP2_CONNECTION_PREFACE)) {
+    ENVOY_LOG(trace, "http inspector: http2 connection preface found");
+    protocol_ = "HTTP/2";
+    done(true);
+  } else {
+    unsigned int id, sid = 0, spaces[2];
```

Comment:
I think you can find only `\r` for end of request line. I think use data.find, then data.substr, then use absl::StrSplit will be clearer what you do here?

Reply:
So I think for inspector, search for `\r` or `\n` is enough.

Reply:
We can easily extract request line from the message by searching for `\r` or `\n` but we loss the chance to verify request line by checking `CRLF` at the end of the request line.

Reply:
Sorry I don't follow, what do we loss? find either `\r` or `\n` is enough since the inspector doesn't care about the second line.

Reply:
Sorry. I mean checking `CRLF` might also help to detect packets that are not http. Not sure if this happens in real world `GET /index HTTP/1.1LFhost:aaa...`.

Reply:
Per discuss offline, will make the change to search `\r` or `\n`.